include/ddc/for_each.hpp

```diff
@@ -38,69 +57,55 @@ struct serial_policy
  * @param[in] f      a functor taking an index as parameter
  */
 template <class... DDims, class Functor>
 inline void for_each_n(serial_policy, DiscreteVector<DDims...> const& extent, Functor&& f) noexcept
 {
-    detail::for_each_impl(
+    detail::for_each_serial(
             DiscreteVector<DDims...> {detail::type_constant_v<DDims, std::ptrdiff_t, 0>...},
             DiscreteVector<DDims...> {get<DDims>(extent) - 1 ...},
             std::forward<Functor>(f));
 }
 
-/** OpenMP parallel execution
- * - default scheduling
- * - collapsing in case of tightly nested loops
- */
+/// OpenMP parallel execution on the outer loop with default scheduling
 struct omp_policy
 {
 };
 
-/** iterates over a 1d domain
- * @param[in] domain the 1d domain to iterate
+/** iterates over a nD domain using the OpenMP execution policy
+ * @param[in] domain the domain over which to iterate
  * @param[in] f      a functor taking an index as parameter
  */
-template <class DDim, class Functor>
-inline void for_each(omp_policy, DiscreteDomain<DDim> const& domain, Functor&& f) noexcept
+template <class... DDims, class Functor>
+inline void for_each(omp_policy, DiscreteDomain<DDims...> const& domain, Functor&& f) noexcept
 {
-    DiscreteDomainIterator<DDim> const it_b = domain.begin();
-    DiscreteDomainIterator<DDim> const it_e = domain.end();
-#pragma omp parallel for default(none) shared(it_b, it_e, f)
-    for (DiscreteDomainIterator<DDim> it = it_b; it != it_e; ++it) {
-        f(*it);
-    }
+    detail::for_each_omp(domain.front(), domain.back(), std::forward<Functor>(f));
 }
 
-/** iterates over a 2d domain
- * @param[in] domain the 2d domain to iterate
- * @param[in] f      a functor taking 2 indices as parameter
+/** iterates over a nD extent using the OpenMP execution policy
+ * @param[in] extent the extent over which to iterate
+ * @param[in] f      a functor taking an index as parameter
  */
-template <class DDim1, class DDim2, class Functor>
-inline void for_each(omp_policy, DiscreteDomain<DDim1, DDim2> const& domain, Functor&& f) noexcept
+template <class... DDims, class Functor>
+inline void for_each_n(omp_policy, DiscreteVector<DDims...> const& extent, Functor&& f) noexcept
 {
-    DiscreteDomainIterator<DDim1> const it1_b = select<DDim1>(domain).begin();
-    DiscreteDomainIterator<DDim1> const it1_e = select<DDim1>(domain).end();
-    DiscreteDomainIterator<DDim2> const it2_b = select<DDim2>(domain).begin();
-    DiscreteDomainIterator<DDim2> const it2_e = select<DDim2>(domain).end();
-#pragma omp parallel for collapse(2) default(none) shared(it1_b, it1_e, it2_b, it2_e, f)
-    for (DiscreteDomainIterator<DDim1> it1 = it1_b; it1 != it1_e; ++it1) {
-        for (DiscreteDomainIterator<DDim2> it2 = it2_b; it2 != it2_e; ++it2) {
-            f(DiscreteCoordinate<DDim1, DDim2>(*it1, *it2));
-        }
-    }
+    detail::for_each_omp(
+            DiscreteVector<DDims...> {detail::type_constant_v<DDims, std::ptrdiff_t, 0>...},
```

Comment:
I committed that without asking you. Do you see any better way to implement it?

Reply:
You mean the null vector ?

Reply:
yep, or more exactly repeat a constant n times

Reply:
there's no `std::repeat_v<sizeof...(DDims), 0>` or equivalent ?

Reply:
I have an other way, better will be quite subjective:
```
{std::integral_constant<std::conditional_t<true, std::ptrdiff_t, std::ptrdiff_t>, 0>::value...}
```

Reply:
I don't think so